Revert harnesses to v0.3.0 and add UTF-8 encoding #74 #72

cirulls opened the pull request:
### Summary

This pull request fixes the Saxon XProc test harness (#74) and integrates the changes made in #72 by @opax for displaying UTF-8 characters in the HTML report.  

### What has changed and why

As identified in #74 by @AirQuick, the Saxon XProc harness modified in release v0.4.0-RC did not work any more whereas the version implemented in Oxygen (based on v0.3.0) is working. Therefore I reverted all the harnesses included in ```src/harnesses``` to [v0.3.0](https://github.com/xspec/xspec/tree/77213c795263f88c6e46ed52ac33a42f057878d0/src/harnesses) and removed files which were not present in v0.3.0. Note that I only tested the Saxon XProc harness.

I added the UTF-8 encoding in the report for all the harnesses as requested in #72.

I included a bats test that checks that the Saxon XProc harness runs correctly and that the HTML report includes  the text ```charset=UTF-8``` (note the ```grep``` command at the end):

```bash
run bash -c "java -Xmx1024m -cp ${XMLCALABASH_CP} com.xmlcalabash.drivers.Main -isource=xspec-72.xspec xspec-home=file:///home/travis/build/xspec/xspec/ -oresult=/home/travis/build/xspec/xspec/test/xspec/xspec-72-report.html /home/travis/build/xspec/xspec/src/harnesses/saxon/saxon-xslt-harness.xproc && grep 'charset=UTF-8' /home/travis/build/xspec/xspec/test/xspec/xspec-72-report.html"
```

I made sure that this test failed with the implementation that didn't specify the encoding. 

I added the XSpec test ```xspec-72.xspec``` but this is not strictly necessary, any XSpec test will do as the bats test is simply testing the presence of ```charset=UTF-8``` in the HTML report. However, I thought it is useful to link this test to #72 so that if it fails the ticket should provide some background to debug it.
 
I added the installation of [XML Calabash](http://xmlcalabash.com/) in the Travis configuration. This is needed to run the bats test that executes the Saxon XProc harness. 

I wrote some documentation on [how to run XSpec with XProc](https://github.com/xspec/xspec/wiki/Running-with-XProc) in the wiki.

### How to review it
Feel free to follow the instructions in the wiki to check that the Saxon XProc harness is working correctly and the documentation is good enough.
 
@opax: could you please check that this pull request includes your changes in #72 and you're happy with it?
@AirQuick : could you please review the documentation in the wiki (I'm not sure about my instructions for Windows) and sanity check my pull request? It would also be good to replicate the Travis configuration and bats test for XProc on AppVeyor, I don't know how feasible this is.

AirQuick commented on .travis.yml:
```diff
@@ -18,6 +18,12 @@ before_script:
+    # install XML Calabash
+    - export XMLCALABASH_VERSION=1.1.15-97
+    - wget -O /tmp/xspec/xmlcalabash-${XMLCALABASH_VERSION}.zip https://github.com/ndw/xmlcalabash1/releases/download/${XMLCALABASH_VERSION}/xmlcalabash-${XMLCALABASH_VERSION}.zip 
+    - sudo apt-get install -y unzip
```
Is this `sudo` line inevitable?
[This article](https://docs.travis-ci.com/user/migrating-from-legacy/) seems to recommend not to use `sudo`.

cirulls replied:
Good point, it's better not to use ```sudo``` - if it can be avoided - for security reasons. I'll try the following things:

1. See if ```unzip``` is already installed by default so there is no need to install it
2. If 1 doesn't work, try to install it without ```sudo```
3. If 2 doesn't work, use the [recommended approach](https://docs.travis-ci.com/user/migrating-from-legacy/#How-do-I-install-APT-sources-and-packages%3F) for container-based infrastructure

cirulls replied:
I removed ```sudo``` as it is not necessary (```unzip``` is already installed by default within Travis).

AirQuick commented on test/xspec.bats:
```diff
@@ -190,3 +190,10 @@ teardown() {
 }
+
+@test "executing the Saxon XProc harness generates a report with UTF-8 encoding" {
+	run bash -c "java -Xmx1024m -cp ${XMLCALABASH_CP} com.xmlcalabash.drivers.Main -isource=xspec-72.xspec xspec-home=file:///home/travis/build/xspec/xspec/ -oresult=/home/travis/build/xspec/xspec/test/xspec/xspec-72-report.html /home/travis/build/xspec/xspec/src/harnesses/saxon/saxon-xslt-harness.xproc && grep 'charset=UTF-8' /home/travis/build/xspec/xspec/test/xspec/xspec-72-report.html"
```
We have to avoid hardcoding the path.
When the repository is forked, the path is different from `xspec/xspec`. The test always fails there as follows.

```shell
not ok 20 executing the Saxon XProc harness generates a report with UTF-8 encoding
# (in test file xspec.bats, line 198)
#   `[ "$status" -eq 0 ]' failed
# 1
# ERROR: err:XD0011:XProc error err:XD0011 ERROR: cause: I/O error reported by XML parser processing file:/home/travis/build/xspec/xspec/src/harnesses/saxon/saxon-xslt-harness.xproc: /home/travis/build/xspec/xspec/src/harnesses/saxon/saxon-xslt-harness.xproc (No such file or directory) ERROR: It is a dynamic error if the resource referenced by a p:document element does not exist, cannot be accessed, or is not a well-formed XML document. ERROR: Underlying exception: I/O error reported by XML parser processing file:/home/travis/build/xspec/xspec/src/harnesses/saxon/saxon-xslt-harness.xproc: /home/travis/build/xspec/xspec/src/harnesses/saxon/saxon-xslt-harness.xproc (No such file or directory)
```
Also [this article](https://docs.travis-ci.com/user/environment-variables/#Default-Environment-Variables) denotes `/home/travis` as `(do not depend on this value)`.

AirQuick replied:
Maybe we can use `file://${TRAVIS_BUILD_DIR}` for `xspec-home`, and the relative path for the rest.

cirulls replied:
I agree, I should not have hardcoded the paths. I'll try to use ```${TRAVIS_BUILD_DIR}``` instead and check the other paths.

AirQuick replied:
Since the report is XHTML and Saxon is readily available, we might want to have Saxon do `grep`'s filtering job and inspect its output in Bats. For example,

```winbatch
P:\xspec>bin\xspec.bat tutorial\escape-for-regex.xspec
...
passed: 5 / pending: 0 / failed: 1 / total: 6
Report available at P:\xspec\tutorial\xspec\escape-for-regex-result.html
Done.

P:\xspec>java -cp "%SAXON_CP%" net.sf.saxon.Query -s:"tutorial\xspec\escape-for-regex-result.html" -qs:"declare default element namespace 'http://www.w3.org/1999/xhtml'; string(html/head/meta[@http-equiv eq 'Content-Type']/@content)" !method=text
text/html; charset=UTF-8
```

cirulls replied:
Yes, I think that's an alternative approach. However, you would still need to chain the two commands (I used ```&&``` in the bats test) or execute them within the same test as the test must check that the HTML contains ```charset=UTF-8```. 

I feel that that the ```grep``` command is shorter and is also very common in Linux. I saw that in [xspec-bat.cmd](https://github.com/xspec/xspec/blob/master/test/xspec-bat.cmd) you used ```findstr``` which I think is the equivalent of ```grep``` in Windows. Would that work for implementing this test?

Let me know if I misunderstood your comment.

AirQuick replied:
Agreed in that aspect. But I'm concerned about the preciseness. Suppose someday someone finds this line

```xml
<x:expect label="the encoding is correct"> 
```

and wonders what the correct means actually. After glancing at the bug record, he would possibly augment the label as follows

```xml
<x:expect label="the encoding is correct; html header has charset=UTF-8">
```

and then the test by `grep` will start to always succeed...

cirulls replied:
I used ```${TRAVIS_BUILD_DIR}``` instead of hardcoding the path.

cirulls replied:
I see your point now. I implemented a bats test that uses Saxon to check the encoding as follows:
```
java -cp ${SAXON_CP} net.sf.saxon.Query -s:${TRAVIS_BUILD_DIR}/test/xspec/xspec-72-result.html -qs:"declare default element namespace 'http://www.w3.org/1999/xhtml'; string(html/head/meta[@http-equiv eq 'Content-Type']/@content) = 'text/html; charset=UTF-8'"
```

An alternative approach to make a more precise bats test with ```grep``` is to use  ```sed``` to select a specific line of the HTML report (line 3 in this case) and pipe it through ```grep``` in order to match the full ```<meta>``` element (not just ```charset=UTF-8```). I report it here for documentation but I have not included it in the latest changes:
```bash
run bash -c "java -Xmx1024m -cp ${XMLCALABASH_CP} com.xmlcalabash.drivers.Main -isource=xspec-72.xspec xspec-home=file://${TRAVIS_BUILD_DIR}/ -oresult=${TRAVIS_BUILD_DIR}/test/xspec/xspec-72-result.html ${TRAVIS_BUILD_DIR}/src/harnesses/saxon/saxon-xslt-harness.xproc && sed '3!d' ${TRAVIS_BUILD_DIR}/test/xspec/xspec-72-result.html | grep '<meta http-equiv=\"Content-Type\" content=\"text/html; charset=UTF-8\" />' "
```

cirulls commented:
Thanks for your review. I'll make the changes you suggested at the beginning of next week and ask again for your approval.

AirQuick commented:
I reviewed [the documentation](https://github.com/xspec/xspec/wiki/Running-with-XProc) and made some edits.

cirulls commented:
Following your review, I implemented the suggested changes, please check my replies to your comments. 

In a nutshell, my latest changes do the following:

1. Remove ```sudo``` from ```.travis``` configuration
2. Use ```TRAVIS_BUILD_PATH``` environment variable instead of hardcoding the path
3. Use Saxon to check the encoding in the bats test 

@AirQuick: Could you please review these changes and see if they address your requests? Thanks.

AirQuick commented on src/harnesses/basex/basex-xquery-harness.xproc:
```diff
@@ -0,0 +1,136 @@
+	<p:option name="compiled-file" select="'file:/tmp/xspec-basex-compiled-suite.xq'"/>
+	<!-- TODO: Use the absolute URIs through the EXPath Packaging System. -->
+
+	<p:variable name="compiler" select="        resolve-uri('src/compiler/generate-query-tests.xsl', $xspec-home)"/>
```
Just a nitpick. 😺 
You seem to have reformatted the XML. So now there are extra spaces, which looks a bit odd. (Spaces between `select="` and `resolve-uri`)

cirulls replied:
It's again a good point. My text editor of choice (vim) doesn't recognise by default ```.xproc``` and ```.xspec``` file so every time I have to remember to reformat it properly. This is why some files I worked on are sometimes inconsistently formatted. I plan to automatically format files against a specific conventions (e.g. 4 spaces for indentation) and set up some styling conventions guidelines (at least for myself) before I do a commit but I haven't got there yet. 

For now I will just remove that extra space and check that this does not appear on other committed files. I'll do this probably tomorrow.

AirQuick commented:
Maybe we can omit the xml declaration from the bats expectation. I tried it on this branch for our convenience. I also followed your best practice to simulate [the failing test](https://travis-ci.org/xspec/xspec/builds/201751243).
What do you think?

cirulls commented:
> Maybe we can omit the xml declaration from the bats expectation.

Yes, that's neater, we don't need the XML declaration. I'll implement this in my branch tomorrow and commit the changes so that you can check them.

cirulls commented:
I removed the useless whitespace for several harnesses files and fixed my previous odd formatting. I'm fine with omitting  the XML declaration, thanks for submitting it directly to my branch. 

Can you review the changes and approve it or request further changes?

cirulls commented:
Thanks for thorough review. I asked @opax to confirm so that we can merge this pull request into master.

cirulls commented:
@opax also [confirmed](https://github.com/xspec/xspec/pull/72#issuecomment-280393610) it so I merge branch 72 into master.
